### PR TITLE
Introduce `TypePlanner` for customizing type planning

### LINKED ISSUE #13292: Allow to customize `convert_data_type` behavior when SQL planning
### Is your feature request related to a problem or challenge?

In `datafusion-sqlparser-rs`, we define many SQL types but they aren't all required by the DataFusion core.
https://github.com/apache/datafusion/blob/b0b6e4427fd6bd64fc29e521c43d86206d2e4ea7/datafusion/sql/src/planner.rs#L524-L529

I think it's good to provide a way to extend or customize the behavior for SQL type to DataFusion type (maybe changed to `NativeType` in the future) for the downstream project that may want to use those unsupported SQL types.

### Describe the solution you'd like

Introduce a new API for `ExprPlanner`,
```
    /// Plan SQL type to DataFusion data type
    ///
    /// Returns None if not possible
    fn plan_data_type(&self, _sql_type: &ast::DataType) -> Option<DataType> {
        None
    }
```

Try planning the SQL type using `ExprPlanner` first. If not present, revert to the default behavior.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #13292 .

## Are these changes tested?
yes

## Are there any user-facing changes?
A new trait for ExprPlanner

### CODE REVIEW

File: datafusion/expr/src/planner.rs [hunk 1]
```diff
@@ -205,6 +206,13 @@ pub trait ExprPlanner: Debug + Send + Sync {
     fn plan_any(&self, expr: RawBinaryExpr) -> Result<PlannerResult<RawBinaryExpr>> {
         Ok(PlannerResult::Original(expr))
     }
+
+    /// Plan SQL type to DataFusion data type
+    ///
+    /// Returns None if not possible
+    fn plan_data_type(&self, _sql_type: &ast::DataType) -> Result<Option<DataType>> {
```
Comment (alamb) [inline datafusion/expr/src/planner.rs:213]: 👍

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -401,6 +401,14 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
     }
 
     pub(crate) fn convert_data_type(&self, sql_type: &SQLDataType) -> Result<DataType> {
+        // First check if any of the registered expr_planners can handle this type
+        for expr_planner in self.context_provider.get_expr_planners() {
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:405]: It seems strange that the planner for individual expressions would handle query-wide data types (though I realize this is currently the only hook that is available into the sql planner)

What would you think about adding a `DataTypePlanner` trait / field to the planner explicitly for this conversion?

Comment (goldmedal) [reply datafusion/sql/src/planner.rs:405]: An independent trait looks good to me.  I'm thinking what's a good way to register the custom planner.
Maybe introduce a method in `ContextProvder::get_data_type_planner`?

```rust
pub trait ContextProvider {
...
  fn get_data_type(&self) -> Option<Arc<dyn DataTypePlanner>>
      None
  }
}
```

We can allow registering the data type planner in `SessionState`. WDYT?

**Review by alamb — APPROVED** [review]
Thank you @goldmedal  -- this looks great

I apologize for the delay in reviewing. I have been away for a few days.

**Comment by goldmedal** [comment]
> Thank you @goldmedal -- this looks great
> 
> I apologize for the delay in reviewing. I have been away for a few days.

Never mind. Thanks @alamb for reviewing.

**Pull request merged**